core/src/main/java/io/ddf/DDFCache.java

```diff
@@ -161,46 +158,56 @@ public DDF getDDF(UUID uuid) throws DDFException {
   }
 
   public DDF getDDFByName(String name) throws DDFException {
-    for(DDF ddf: this.listDDFs()) {
-      if(!Strings.isNullOrEmpty(ddf.getName()) && ddf.getName().equals(name)) {
+    for (DDF ddf : this.listDDFs()) {
+      if (!Strings.isNullOrEmpty(ddf.getName()) && ddf.getName().equals(name)) {
         return ddf;
       }
     }
     throw new DDFException(String.format("Cannot find ddf with name %s", name));
   }
 
-  public synchronized void setDDFName(DDF ddf, String name) throws DDFException {
-    synchronized (ddf) {
-      if (!Strings.isNullOrEmpty(name)) {
-        if (!Strings.isNullOrEmpty(ddf.getName())) {
-          this.mUris.remove(ddf.getUri());
+  public void setDDFName(DDF ddf, String name) throws DDFException {
+    Lock lock = uuidLocks.get(ddf.getUUID());
+    try {
+      if(lock.tryLock(120, TimeUnit.SECONDS)) {
+        try {
+          if (!Strings.isNullOrEmpty(name)) {
+            if (!Strings.isNullOrEmpty(ddf.getName())) {
+              this.mUris.remove(ddf.getUri());
+            }
+            ddf.setName(name);
+            mDDFCache.invalidate(ddf.getUUID());
+            mDDFwithNameCache.put(ddf.getUUID(), ddf);
+            this.mUris.put(ddf.getUri(), ddf.getUUID());
+          } else {
+            throw new DDFException(String.format("DDF's name cannot be null or empty"));
+          }
+        } finally {
+          lock.unlock();
         }
-        ddf.setName(name);
-
-        mDDFCache.invalidate(ddf.getUUID());
-        mDDFwithNameCache.put(ddf.getUUID(), ddf);
-        this.mUris.put(ddf.getUri(), ddf.getUUID());
       } else {
-        throw new DDFException(String.format("DDF's name cannot be null or empty"));
+        throw new DDFException("Timeout setting DDF name");
       }
+    } catch (Exception e) {
+      throw new DDFException(e);
     }
   }
 
   public synchronized void setDDFUUID(DDF ddf, UUID uuid) throws DDFException {
-    if(this.containsDDF(uuid)) {
+    if (this.containsDDF(uuid)) {
       throw new DDFException(String.format("DDF with uuid %s already exists", uuid));
     } else {
       UUID prevUUID = ddf.getUUID();
       ddf.setUUID(uuid);
-      if(prevUUID != null) {
-        if(ddf.getName() == null) {
+      if (prevUUID != null) {
+        if (ddf.getName() == null) {
           mDDFCache.invalidate(prevUUID);
```

Comment:
why do we need to invalidate prevUUID here ?
as this time, UUID isn't in your cache anymore.